src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -152,121 +91,199 @@ def __init__(
+        # processes without directly sharing memory state, which makes it safe to use in a spawn context.
+        self._input_queues = [manager.Queue() for _ in range(self._n_process)]
+        self._output_queues = [manager.Queue() for _ in range(self._n_process)]
+        self._control_signal_queue = manager.Queue()
+        self._process_info = manager.dict()
+
+        # when using fork, we first create a process with spawn method to establish a clean environment
+        # Then fork the subprocess in this environment to avoid some deadlock problems
+        if self._use_fork:
+            # 1. Create input_queue, output_queue, control_signal_queue and _process_info in the main process.
+            # 2. Pass the above queue/dict as parameters to spawn and fork processes to transfer information
+            # between processes.
+            self._processes_manager = ForkProcessManager(
+                self._control_signal_queue,
+                self._flow_file,
+                self._connections,
+                self._working_dir,
+                self._input_queues,
+                self._output_queues,
+                self._process_info,
+                _process_wrapper,
+                False,
+            )
+        else:
+            executor_creation_func = partial(
+                FlowExecutor.create,
+                flow_file=self._flow_file,
+                connections=self._connections,
+                working_dir=self._working_dir,
+                raise_ex=False,
+            )
+
+            # 1. Create input_queue, output_queue, and _process_info in the main process.
+            # 2. Spawn _n_process sub-process and pass the above queue/dict to these sub-process to transfer information
+            # between main process and sub process.
+            self._processes_manager = SpawnProcessManager(
+                executor_creation_func,
+                self._input_queues,
+                self._output_queues,
+                self._process_info,
+                _process_wrapper,
+                False,
+            )
+
+        self._processes_manager.start_processes()
+
+        monitor_pool = ThreadPool(self._n_process, initializer=set_context, initargs=(contextvars.copy_context(),))
+        self._monitor_pool = monitor_pool
 
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        if self._pool is not None:
-            self._pool.close()
-            self._pool.join()
+        if self._monitor_pool is not None:
+            self._monitor_pool.close()
+            self._monitor_pool.join()
+
+    def _get_process_info(self, index):
+        while True:
+            try:
+                process_id = self._process_info[index].process_id
+                process_name = self._process_info[index].process_name
+                return (index, process_id, process_name)
+            except KeyError:
+                continue
```

Comment:
Explain the continue here

Reply:
fixed.

Reply:
It seems to be an infinite loop if it is always not ready, we need to sleep and wait for a while.

Reply:
fixed